Fixed level check for pet spells

DotesMiqotes opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

<!-- Describe what your PR does here. If it closes an existing issue, you can mention: "Closes #1234" and GitHub will link this PR to that issue. -->
Fixes issue #5318 by changing the level check for pets casting spells to check the `min_level` value from `mob_spell_lists.sql` instead of checking if a player of the pet's job and level can cast the spell.
Fully optimizes `CanUseSpell()` by converting `PCaster->objtype` if statements to `switch(PCaster->objtype)`, removing no longer needed variables, and moving all function calls to their relevant cases.

## Steps to test these changes

<!-- Clear and detailed steps to test your changes here -->
Summoned Light Spirit at various levels in Cape Teriggan and let it go through it's spell list multiple times while fighting things, before and after recompiling the server with the changes. Before it would stop casting Banish after level 90, and it was able to cast Banish IV (non-player spell on it's spell list) after.
Tested every possible case I could think of covered by the optimization to ensure everything is working properly.

DotesMiqotes committed:
Fixed level check for pet spells

Changes the level check for pets casting spells to check the `min_level` value from `mob_spell_lists.sql` instead of checking if a player of the same job can cast the spell.

zach2good commented on src/map/spell.cpp:
```diff
@@ -687,7 +687,7 @@ namespace spell
             uint8 requirements = spell->getRequirements();
 
-            if (PCaster->objtype == TYPE_MOB || (PCaster->objtype == TYPE_PET && static_cast<CPetEntity*>(PCaster)->getPetType() == PET_TYPE::AUTOMATON) ||
+            if (PCaster->objtype == TYPE_MOB || (PCaster->objtype == TYPE_PET && static_cast<CPetEntity*>(PCaster)->getPetType() != PET_TYPE::ADVENTURING_FELLOW) ||
```
Original logic:
- If the caster if a mob, or a pet automaton, or a trust

New logic:
- If the caster is a mob, or a pet that isn't an adventuring fellow, or a trust

Does this drag in and apply changes to the other pet types?
```cpp
enum class PET_TYPE : uint8
{
    AVATAR             = 0,
    WYVERN             = 1,
    JUG_PET            = 2,
    CHARMED_MOB        = 3,
    AUTOMATON          = 4,
    ADVENTURING_FELLOW = 5,
    CHOCOBO            = 6,
    LUOPAN             = 7,
};
```

DotesMiqotes replied:
It does, it's also why I excluded adventuring fellows in particular. From what I could see they were the only pet type that can cast spells but didn't have a spell list in mob_spell_lists.sql.

zach2good commented on src/map/spell.cpp:
```diff
@@ -696,8 +696,9 @@ namespace spell
-                // ensure trust level is appropriate+
-                if (PCaster->objtype == TYPE_TRUST && PCaster->GetMLevel() < static_cast<CMobEntity*>(PCaster)->m_SpellListContainer->GetSpellMinLevel(spell->getID()))
+                // Ensure pet or trust is level appropriate
+                if (((PCaster->objtype == TYPE_PET && static_cast<CPetEntity*>(PCaster)->getPetType() != PET_TYPE::AUTOMATON) || PCaster->objtype == TYPE_TRUST) &&
```
Same down here, the logic is getting very crowded and hard to reason about. I'd recommend breaking a bunch of these out into more human readable flags at the top of this block (near `uint8 requirements = spell->getRequirements();`:
```cpp
const bool isPet = PCaster->objtype == TYPE_PET;
const bool isMob = PCaster->objtype == TYPE_MOB;
const bool isTrust =  PCaster->objtype == TYPE_TRUST;
const bool isPetButNotFellow = isPet  && static_cast<CPetEntity*>(PCaster)->getPetType() != PET_TYPE::ADVENTURING_FELLOW;
```

and then use those, stacked on different lines:
```cpp
if (isMob ||
    isPetButNotFellow ||
    isTrust)
{
    ....
```

DotesMiqotes replied:
Sounds good, I'll add these.

DotesMiqotes replied:
I have a copy of `spell.cpp` with those changes added and working. While I was looking over the commit, I realized that I could handle the different `objtype` cases with a switch. I got a second `spell.cpp` working with the switch statement, the change is drastic so I wanted to clear it with you before making the commit.
```c++
// Check If user can cast spell
bool CanUseSpell(CBattleEntity* PCaster, CSpell* spell)
{
    if (spell == nullptr)
    {
        return false;
    }

    bool usable = false;
    uint8 requirements = spell->getRequirements();

    switch(PCaster->objtype)
    {
        case TYPE_MOB:
            // Unable to cast because caster is hidden or untargetable
            if (PCaster->IsNameHidden() || static_cast<CMobEntity*>(PCaster)->GetUntargetable())
            {
                return false;
            }
            // Mobs can cast any non-given char spell
            return true;

        case TYPE_PC:
            if (spell->getSpellGroup() == SPELLGROUP_TRUST)
            {
                return true; // every PC can use trusts
            }
            else if (luautils::OnCanUseSpell(PCaster, spell))
            {
                return true;
            }
            [[fallthrough]];
        case TYPE_FELLOW:
        case TYPE_NPC:
        case TYPE_SHIP: // This can be removed it does technically follow current logic though
            // Make sure caster has the right main job and level
            if (PCaster->GetMLevel() >= spell->getJob(PCaster->GetMJob()))
```
Removed what goes between here because it's the current code.
```c++
        case TYPE_PET:
        case TYPE_TRUST:
            // Unable to cast because caster is hidden or untargetable
            if (PCaster->IsNameHidden() || static_cast<CMobEntity*>(PCaster)->GetUntargetable())
            {
                return false;
            }

            if (static_cast<CPetEntity*>(PCaster)->getPetType() == PET_TYPE::AUTOMATON)
            {
                return true;
            }

            // Ensure pet or trust is level appropriate
            if (PCaster->GetMLevel() < static_cast<CMobEntity*>(PCaster)->m_SpellListContainer->GetSpellMinLevel(spell->getID()))
            {
                return false;
            }
            return true;
```

zach2good commented:
Please look into how to use `git commit --amend` and `git push --force`, or `git rebase` to squash commits. 

`Removed 2 spaces in spell.cpp` isn't a useful commit and should be rolled into your original commit.

https://github.com/LandSandBoat/server/wiki/Development-Guide-Git#writing-useful-commit-messages

DotesMiqotes commented:
Understood, I'll go ahead and squash those into one commit, and use `git commit --amend` in the future for those minor formatting issues.

DotesMiqotes committed:
Fully Optimized CanUseSpell()

Fully optimized `CanUseSpell()` by converting `PCaster->objtype` if statements to `switch(PCaster->objtype)`, removing no longer needed variables, and moving all function calls to their relevant cases.
